File: kotlin/src/main/kotlin/io/github/ocirne/aoc/year2024/Day11.kt
package io.github.ocirne.aoc.year2024

import io.github.ocirne.aoc.AocChallenge
import kotlin.math.log10
import kotlin.math.pow

class Day11(val lines: List<String>) : AocChallenge(2024, 11) {

    class Blink(line: String) {

        private val startNumbers = line.trim().split(' ').map { it.toLong() }

        private val cache = mutableMapOf<Pair<Long, Int>, Long>()

        private fun recursiveBlink(x: Long, step: Int): Long {
            if (x to step in cache) {
                return cache[x to step]!!
            }
            if (step == 0) {
                return 1
            }
            val length = log10(x.toDouble()).toInt() + 1
            val result = when {
                x == 0L -> recursiveBlink(1, step - 1)
                length % 2 == 0 -> {
                    val m = (10.0).pow(length / 2).toInt()
                    recursiveBlink(x / m, step - 1) + recursiveBlink(x % m, step - 1)
                }
                else -> recursiveBlink(2024 * x, step - 1)
            }
            cache[x to step] = result
            return result
        }

        fun blinks(times: Int): Long {
            return startNumbers.sumOf { recursiveBlink(it, times) }
        }
    }

    override fun part1(): Long {
        return Blink(lines.first()).blinks(25)
    }

    override fun part2(): Long {
        return Blink(lines.first()).blinks(75)
    }
}

File: python/src/aoc/year2024/day11.py
from aoc.util import load_input, load_example

from math import log10
from functools import lru_cache


@lru_cache(maxsize=None)
def recursive_blink(s, d):
    if d == 0:
        return 1
    if s == 0:
        return recursive_blink(1, d - 1)
    hl, r = divmod(int(log10(s)) + 1, 2)
    if r == 0:
        first, second = divmod(s, 10**hl)
        return recursive_blink(first, d - 1) + recursive_blink(second, d - 1)
    return recursive_blink(2024 * s, d - 1)


def bar(line, blinks):
    return sum(recursive_blink(int(s), blinks) for s in line.split())


def part1(lines):
    """
    >>> part1(load_example(__file__, "11"))
    55312
    """
    return bar(lines[0], 25)


def part2(lines):
    return bar(lines[0], 75)


if __name__ == "__main__":
    data = load_input(__file__, 2024, "11")
    print(part1(data))
    print(part2(data))
